rust/relay/src/sockets.rs

```diff
@@ -0,0 +1,259 @@
+                        port,
+                        from,
+                        packet: &self.buffer[..num_bytes],
+                    }));
+                }
+            }
+
+            match ready!(self.event_rx.poll_recv(cx)) {
+                Some(Event::NewSocket(token, socket)) => {
+                    self.inner.insert(token, socket);
+                    continue;
+                }
+                Some(Event::SocketReady(ready)) => {
+                    self.current_ready_socket = Some(ready);
+                    continue;
+                }
+                Some(Event::Crashed(error)) => {
+                    return Poll::Ready(Err(Error::MioTaskCrashed(error)));
+                }
+                None => {
+                    panic!("must not poll `Sockets` after mio task exited")
+                }
+            };
+        }
+    }
+}
+
+/// A packet read from a socket.
+#[derive(Debug)]
+pub struct Received<'a> {
+    pub port: u16,
+    pub from: SocketAddr,
+    pub packet: &'a [u8],
+}
+
+#[derive(Debug)]
+pub enum Error {
+    Io(io::Error),
+    MioTaskCrashed(anyhow::Error),
+}
+
+enum Command {
+    NewSocket((u16, AddressFamily)),
+    DisposeSocket(mio::net::UdpSocket),
+}
+
+enum Event {
+    NewSocket(mio::Token, mio::net::UdpSocket),
+    SocketReady(mio::Token),
+    Crashed(anyhow::Error),
+}
+
+fn not_connected(port: u16, address_family: AddressFamily) -> io::Error {
+    io::Error::new(
+        io::ErrorKind::NotConnected,
+        format!("No socket for port {port} on address family {address_family}"),
+    )
+}
+
+fn mio_worker_task(
+    event_tx: mpsc::Sender<Event>,
+    mut cmd_rx: mpsc::UnboundedReceiver<Command>,
+) -> Result<()> {
+    let mut poll = mio::Poll::new()?;
+    let mut events = mio::Events::with_capacity(1024);
+
+    loop {
+        poll.poll(&mut events, Some(Duration::from_secs(1)))?;
+
+        for event in events.iter() {
+            event_tx.blocking_send(Event::SocketReady(event.token()))?;
+        }
+
+        match cmd_rx.try_recv() {
+            Ok(Command::NewSocket((port, af))) => {
+                let mut socket = mio::net::UdpSocket::from_std(make_wildcard_socket(af, port)?);
+                let token = token_from_port_and_address_family(port, af);
+
+                poll.registry()
+                    .register(&mut socket, token, mio::Interest::READABLE)?;
+
+                event_tx.blocking_send(Event::NewSocket(token, socket))?;
+            }
+            Ok(Command::DisposeSocket(mut socket)) => {
+                poll.registry().deregister(&mut socket)?;
+            }
+            Err(mpsc::error::TryRecvError::Empty) => {}
+            Err(mpsc::error::TryRecvError::Disconnected) => {
+                bail!("Command channel disconnected")
+            }
+        }
+    }
+}
+
+/// Encodes a port (u16) and an [`AddressFamily`] into an [`mio::Token`] by flipping the 17th bit of the internal [`usize`] based on the [`AddressFamily`].
```

Comment:
So we make up the tokens, and then Mio treats them as opaque, and they're always `usize`

At first I had guessed Mio was generating the tokens.

Reply:
> So we make up the tokens, and then Mio treats them as opaque, and they're always usize

Correct.